shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -256,18 +256,17 @@ class RequirementSet {
   }
 }
 
-const permissionIdPattern = Match.OneOf({ canAccess: null }, { appDefined: Match.Integer });
-// As we compute a flow of permissions, we need to be able to explicitly refer to not only
-// the permissions enumerated in a grain's `ViewInfo`, but also the usually-implicit "can
-// access the grain at all" permission. A permission ID, defined here, allows us to refer
-// to either of these types of permissions.
+// A permission ID is a stringified integer greater than or equal to -1. The permission "-1"
```

Comment:
> keys in plain javascript objects are always strings

Sort of but not exactly. An array in Javascript is actually an object:

    > var x = ["foo", "bar"]
    > typeof x
    'object'
    > Object.keys(x)
    [ '0', '1' ]
    > x.baz = "qux"
    > x
    [ 'foo', 'bar', baz: 'qux' ]
    > Object.keys(x)
    [ '0', '1', 'baz' ]

Similarly, objects can have numeric keys:

    > ({0: "foo", 1: "bar"})[1]
    'bar'

Theoretically a non-string key is converted to a string for all purposes, but that's not actually what happens under the hood. v8 treats numeric keys (including string keys that happen to be canonical representations of integers) differently from strings and optimizes for them. Internally, the difference between an object and an array is actually pretty fuzzy.

But in any case, it is perfectly fine to use numbers as keys, or use an array with extra non-numeric fields.

> -1 because that makes it somewhat easier to iterate through the permissions and to find the maximum key.

But it's significantly less self-documenting, and if we end up with other pseudo-permissions in the future it gets uglier...